spock-core/src/main/java/org/spockframework/compiler/SpecRewriter.java

```diff
@@ -515,43 +522,56 @@ public Block getCurrentBlock() {
   }
 
   @Override
-  public void defineRecorders(List<Statement> stats, boolean enableErrorCollector, String recorderSuffix) {
+  public void defineValueRecorder(List<Statement> stats, String variableNameSuffix) {
     // recorder variable needs to be defined in outermost scope,
     // hence we insert it at the beginning of the block
-    List<Statement> allStats = new ArrayList<>(stats);
-
-    stats.clear();
-
     stats.add(0,
         new ExpressionStatement(
             new DeclarationExpression(
-                new VariableExpression(SpockNames.VALUE_RECORDER + recorderSuffix, nodeCache.ValueRecorder),
+                new VariableExpression(SpockNames.VALUE_RECORDER + variableNameSuffix, nodeCache.ValueRecorder),
                 Token.newSymbol(Types.ASSIGN, -1, -1),
                 new ConstructorCallExpression(
                     nodeCache.ValueRecorder,
                     ArgumentListExpression.EMPTY_ARGUMENTS))));
+  }
+
+  // This is necessary as otherwise within `with(someMap) { ... }` the variable is resolved to `null`,
+  // but having this local variable beats the resolving against the closure delegate.
+  private void defineErrorRethrower(List<Statement> stats) {
     stats.add(0,
-        new ExpressionStatement(
-            new DeclarationExpression(
-                new VariableExpression(SpockNames.ERROR_COLLECTOR + recorderSuffix, nodeCache.ErrorCollector),
-                Token.newSymbol(Types.ASSIGN, -1, -1),
-                new ConstructorCallExpression(
-                    enableErrorCollector ? nodeCache.ErrorCollector : nodeCache.ErrorRethrower,
-                    ArgumentListExpression.EMPTY_ARGUMENTS))));
+      new ExpressionStatement(
+        new DeclarationExpression(
+          new VariableExpression(SpockNames.ERROR_COLLECTOR, nodeCache.ErrorCollector),
```

Comment:
I think I would prefer `ErrorRethrower.INSTANCE` instead of putting a field in SpecInternals